File: app/src/main/java/fr/free/nrw/commons/contributions/Contribution.java
package fr.free.nrw.commons.contributions;

import android.net.Uri;
import android.os.Parcel;
import androidx.annotation.Nullable;
import androidx.room.Entity;
import fr.free.nrw.commons.Media;
import fr.free.nrw.commons.auth.SessionManager;
import fr.free.nrw.commons.upload.UploadItem;
import fr.free.nrw.commons.upload.UploadMediaDetail;
import fr.free.nrw.commons.upload.WikidataPlace;
import fr.free.nrw.commons.upload.structure.depictions.DepictedItem;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Objects;

@Entity(tableName = "contribution")
public class Contribution extends Media {

    // No need to be bitwise - they're mutually exclusive
    public static final int STATE_COMPLETED = -1;
    public static final int STATE_FAILED = 1;
    public static final int STATE_QUEUED = 2;
    public static final int STATE_IN_PROGRESS = 3;

    private int state;
    private long transferred;
    private String decimalCoords;
    private String dateCreatedSource;
    private WikidataPlace wikidataPlace;
    /**
     * Each depiction loaded in depictions activity is associated with a wikidata entity id, this Id
     * is in turn used to upload depictions to wikibase
     */
    private List<DepictedItem> depictedItems = new ArrayList<>();
    private String mimeType;
    @Nullable
    private Uri localUri;
    private long dataLength;
    private Date dateCreated;

    public Contribution() {
    }

    public Contribution(final UploadItem item, final SessionManager sessionManager,
        final List<DepictedItem> depictedItems, final List<String> categories) {
        super(
            item.getFileName(),
            UploadMediaDetail.formatCaptions(item.getUploadMediaDetails()),
            UploadMediaDetail.formatDescriptions(item.getUploadMediaDetails()),
            sessionManager.getAuthorName(),
            categories);
        localUri = item.getMediaUri();
        decimalCoords = item.getGpsCoords().getDecimalCoords();
        dateCreatedSource = "";
        this.depictedItems = depictedItems;
        wikidataPlace = WikidataPlace.from(item.getPlace());
    }

    public void setDateCreatedSource(final String dateCreatedSource) {
        this.dateCreatedSource = dateCreatedSource;
    }

    public String getDateCreatedSource() {
        return dateCreatedSource;
    }

    public long getTransferred() {
        return transferred;
    }

    public void setTransferred(final long transferred) {
        this.transferred = transferred;
    }

    public int getState() {
        return state;
    }

    public void setState(final int state) {
        this.state = state;
    }

    /**
     * @return array list of entityids for the depictions
     */
    public List<DepictedItem> getDepictedItems() {
        return depictedItems;
    }

    public void setWikidataPlace(final WikidataPlace wikidataPlace) {
        this.wikidataPlace = wikidataPlace;
    }

    public WikidataPlace getWikidataPlace() {
        return wikidataPlace;
    }

    public String getDecimalCoords() {
        return decimalCoords;
    }

    public void setDecimalCoords(final String decimalCoords) {
        this.decimalCoords = decimalCoords;
    }

    public void setDepictedItems(final List<DepictedItem> depictedItems) {
        this.depictedItems = depictedItems;
    }

    public String getMimeType() {
        return mimeType;
    }

    public void setMimeType(final String mimeType) {
        this.mimeType = mimeType;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(final Parcel dest, final int flags) {
        super.writeToParcel(dest, flags);
        dest.writeInt(state);
        dest.writeLong(transferred);
        dest.writeString(decimalCoords);
        dest.writeString(dateCreatedSource);
    }

    /**
     * Constructor that takes Media object and state as parameters and builds a new Contribution object
     * @param media
     * @param state
     */
    public Contribution(Media media, int state) {
        super(media);
        this.state = state;
    }

    protected Contribution(final Parcel in) {
        super(in);
        state = in.readInt();
        transferred = in.readLong();
        decimalCoords = in.readString();
        dateCreatedSource = in.readString();
    }

    public static final Creator<Contribution> CREATOR = new Creator<Contribution>() {
        @Override
        public Contribution createFromParcel(final Parcel source) {
            return new Contribution(source);
        }

        @Override
        public Contribution[] newArray(final int size) {
            return new Contribution[size];
        }
    };

    @Nullable
    public Uri getLocalUri() {
        return localUri;
    }

    public void setLocalUri(@Nullable Uri localUri) {
        this.localUri = localUri;
    }

    public long getDataLength() {
        return dataLength;
    }

    public void setDataLength(long dataLength) {
        this.dataLength = dataLength;
    }

    public Date getDateCreated() {
        return dateCreated;
    }

    public void setDateCreated(Date dateCreated) {
        this.dateCreated = dateCreated;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        if (!super.equals(o)) {
            return false;
        }
        final Contribution that = (Contribution) o;
        return state == that.state &&
            transferred == that.transferred &&
            dataLength == that.dataLength &&
            Objects.equals(decimalCoords, that.decimalCoords) &&
            Objects.equals(dateCreatedSource, that.dateCreatedSource) &&
            Objects.equals(wikidataPlace, that.wikidataPlace) &&
            Objects.equals(depictedItems, that.depictedItems) &&
            Objects.equals(mimeType, that.mimeType) &&
            Objects.equals(localUri, that.localUri) &&
            Objects.equals(dateCreated, that.dateCreated);
    }

    @Override
    public int hashCode() {
        return Objects
            .hash(super.hashCode(), state, transferred, decimalCoords, dateCreatedSource,
                wikidataPlace,
                depictedItems, mimeType, localUri, dataLength, dateCreated);
    }
}

File: app/src/main/java/fr/free/nrw/commons/contributions/Contribution.kt
package fr.free.nrw.commons.contributions

import android.net.Uri
import android.os.Parcelable
import androidx.room.Embedded
import androidx.room.Entity
import androidx.room.PrimaryKey
import fr.free.nrw.commons.CommonsApplication
import fr.free.nrw.commons.Media
import fr.free.nrw.commons.auth.SessionManager
import fr.free.nrw.commons.upload.UploadItem
import fr.free.nrw.commons.upload.UploadMediaDetail
import fr.free.nrw.commons.upload.WikidataPlace
import fr.free.nrw.commons.upload.WikidataPlace.Companion.from
import fr.free.nrw.commons.upload.structure.depictions.DepictedItem
import kotlinx.parcelize.Parcelize
import java.io.File
import java.util.Date

@Entity(tableName = "contribution")
@Parcelize
data class Contribution constructor(
    @Embedded(prefix = "media_") val media: Media,
    @PrimaryKey val pageId: String = media.pageId,
    var state: Int = 0,
    var transferred: Long = 0,
    val decimalCoords: String? = null,
    var dateCreatedSource: String? = null,
    var wikidataPlace: WikidataPlace? = null,
    var chunkInfo: ChunkInfo? = null,
    /**
     * @return array list of entityids for the depictions
     */
    /**
     * Each depiction loaded in depictions activity is associated with a wikidata entity id, this Id
     * is in turn used to upload depictions to wikibase
     */
    val depictedItems: List<DepictedItem> = ArrayList(),
    var mimeType: String? = null,
    val localUri: Uri? = null,
    var dataLength: Long = 0,
    var dateCreated: Date? = null,
    var dateCreatedString: String? = null,
    var dateModified: Date? = null,
    var hasInvalidLocation : Int =  0,
    var contentUri: Uri? = null,
    var countryCode : String? = null,
    var imageSHA1 : String? = null,
    /**
     * Number of times a contribution has been retried after a failure
     */
    var retries: Int = 0
) : Parcelable {

    fun completeWith(media: Media): Contribution {
        return copy(pageId = media.pageId, media = media, state = STATE_COMPLETED)
    }

    constructor(
        item: UploadItem,
        sessionManager: SessionManager,
        depictedItems: List<DepictedItem>,
        categories: List<String>,
        imageSHA1: String
    ) : this(
        Media(
            formatCaptions(item.uploadMediaDetails),
            categories,
            item.fileName,
            formatDescriptions(item.uploadMediaDetails),
            sessionManager.userName,
            sessionManager.userName
        ),
        localUri = item.mediaUri,
        decimalCoords = item.gpsCoords.decimalCoords,
        dateCreatedSource = "",
        depictedItems = depictedItems,
        wikidataPlace = from(item.place),
        contentUri = item.contentUri,
        dateCreatedString = item.fileCreatedDateString,
        imageSHA1 = imageSHA1
    )

    /**
     * Set this true when ImageProcessor has said that the location is invalid
     * @param hasInvalidLocation
     */
    fun setHasInvalidLocation(hasInvalidLocation: Boolean) {
        this.hasInvalidLocation = if (hasInvalidLocation) 1 else 0
    }

    fun hasInvalidLocation(): Boolean {
        return hasInvalidLocation == 1
    }

    companion object {
        const val STATE_COMPLETED = -1
        const val STATE_FAILED = 1
        const val STATE_QUEUED = 2
        const val STATE_IN_PROGRESS = 3
        const val STATE_PAUSED = 4
        const val STATE_QUEUED_LIMITED_CONNECTION_MODE=5

        /**
         * Formatting captions to the Wikibase format for sending labels
         * @param uploadMediaDetails list of media Details
         */
        fun formatCaptions(uploadMediaDetails: List<UploadMediaDetail>) =
            uploadMediaDetails.associate { it.languageCode!! to it.captionText }
                .filter { it.value.isNotBlank() }

        /**
         * Formats the list of descriptions into the format Commons requires for uploads.
         *
         * @param descriptions the list of descriptions, description is ignored if text is null.
         * @return a string with the pattern of {{en|1=descriptionText}}
         */
        fun formatDescriptions(descriptions: List<UploadMediaDetail>) =
            descriptions.filter { it.descriptionText.isNotEmpty() }
                .joinToString(separator = "") { "{{${it.languageCode}|1=${it.descriptionText}}}" }
    }

    val fileKey : String? get() = chunkInfo?.uploadResult?.filekey
    val localUriPath: File? get() = localUri?.path?.let { File(it) }

    fun isCompleted(): Boolean {
        return chunkInfo != null && chunkInfo!!.totalChunks == chunkInfo!!.indexOfNextChunkToUpload
    }

    fun isPaused(): Boolean {
        return CommonsApplication.pauseUploads[pageId] ?: false
    }

    fun unpause() {
        CommonsApplication.pauseUploads[pageId] = false
    }
}
